pkg/lifecycle/lifecycle.go

```diff
@@ -39,58 +40,39 @@ func New(kubeClient clients.ClientsInterface, storage storage.Storage) Lifecycle
 	}
 }
 
-func (l *lifecycle) GetPodFromDaemonSet(key types.NamespacedName) unstructured.UnstructuredList {
-	ds := &unstructured.Unstructured{}
-	ds.SetAPIVersion("apps/v1")
-	ds.SetKind("DaemonSet")
+func (l *lifecycle) GetPodFromDaemonSet(key types.NamespacedName) *v1.PodList {
+	ds := &appsv1.DaemonSet{}
 
 	err := l.kubeClient.Get(context.TODO(), key, ds)
 	if apierrors.IsNotFound(err) || err != nil {
 		warn.OnError(err)
-		return unstructured.UnstructuredList{}
+		return &v1.PodList{}
 	}
 
-	return l.getPodListForUpperObject(ds)
+	return l.getPodListForUpperObject(ds.Spec.Selector.MatchLabels, key.Namespace)
 }
 
-func (l *lifecycle) GetPodFromDeployment(key types.NamespacedName) unstructured.UnstructuredList {
-
-	dp := &unstructured.Unstructured{}
-	dp.SetAPIVersion("apps/v1")
-	dp.SetKind("Deployment")
+func (l *lifecycle) GetPodFromDeployment(key types.NamespacedName) *v1.PodList {
+	dp := &appsv1.Deployment{}
 
 	err := l.kubeClient.Get(context.TODO(), key, dp)
 	if apierrors.IsNotFound(err) || err != nil {
 		warn.OnError(err)
-		return unstructured.UnstructuredList{}
+		return &v1.PodList{}
 	}
 
-	return l.getPodListForUpperObject(dp)
+	return l.getPodListForUpperObject(dp.Spec.Selector.MatchLabels, key.Namespace)
 }
 
-func (l *lifecycle) getPodListForUpperObject(obj *unstructured.Unstructured) unstructured.UnstructuredList {
-	pl := unstructured.UnstructuredList{}
-	pl.SetKind("PodList")
-	pl.SetAPIVersion("v1")
-
-	labels, found, err := unstructured.NestedMap(obj.Object, "spec", "selector", "matchLabels")
-	if err != nil || !found {
-		warn.OnError(err)
-		return pl
-	}
-
-	matchLabels := make(map[string]string)
-	for k, v := range labels {
-		matchLabels[k] = v.(string)
-	}
+func (l *lifecycle) getPodListForUpperObject(matchLabels map[string]string, ns string) *v1.PodList {
```

Comment:
What about moving the client.ListOption into the clients package implementation? clients package is the one that should deal with kuberentes APIs, it can receive just a buch of string parameters, and it will decide if to use client.ListOption struct or anything else? In addition it means that we can just delete getPodListForUpperObject function and call the clients interface directly from the GetPOdsForDeployment/GetPodForDaemonSet/etc' function

Reply:
@yevgeny-shnaidman I am not sure I understand what you are suggesting. Would you make `ClientsInterface.List` accept additional parameters such as the namespace and the matching labels? That does not sound very scalable to me.

Reply:
I am slightly reluctant to make the `ClientsInterface` bigger, but I also see that we filter by namespace in the code. Let's merge this now and have that discussion later.